docs/cpm.rst

```diff
@@ -0,0 +1,253 @@
+
+.. literalinclude:: /packages/packages.rst
+
+rapids-cmake package override
+#############################
+
+rapids-cmake allows projects to override the default values for the pre-defined packages provided by rapids-cmake while also overriding any :cmake:command:`rapids_cpm_find`, `CPM <https://github.com/cpm-cmake/CPM.cmake>`_, and :cmake:module:`FetchContent() <cmake:module:FetchContent>` package calls.
+
+When an override for a project exists no :cmake:module:`find_package()<cmake:command:find_package>` search for that project will occur. This is done to make sure that the requested modified version is used.
+
+If a project is listed in multiple override files, the first file values will be used,
+and all later calls for that packaged will be ignored.  This "first to record, wins"
+approach is used to match FetchContent, and allows parent projects to override child
+projects.
+
+If the override is of an existing default project, it only needs to specify the
+fields it wants to override.
+
+For example if you wanted to change the version of `fmt` you would do the following:
+
+.. literalinclude:: /packages/example_git_tag_override.json
+  :language: json
+
+If you want to specify a completely new project, you need to specify at a minimum all
+the required fields:
+
+.. literalinclude:: /packages/example_new_project_override.json
+  :language: json
+
+
+reproducible rapids-cmake builds
+################################
+
+The rapids-cmake default `versions.json` uses branch names or git tag names
+for dependencies. This is done so that projects can 'live at head' of dependencies.
+This directly conflicts with the concept of reproducible release packages that each
+time they are built contain the exact same byteconde.
+
+:cmake:command:`rapids_cpm_generate_pinned_versions` can be used to generate a set of
+rapids-cmake dependencies from the existing branch names to explicit pinned git SHA1 values. This results in a fully reproducible set of dependencies when building.
+
+To utilize this behavior in your release CI/CD something like the following needs
+to be setup:
+
+  1. Enable generation of a pinned versions during builds via
+  :cmake:command:`rapids_cpm_generate_pinned_versions` or by specifying the :cmake:variable:`RAPIDS_CMAKE_CPM_PINNED_VERSIONS_FILE`.
+  2. If build is good, create the release branch and commit the generated pinned `versions.json` to the repository
+  3. Rebuilds of the project using the pinned version are now possible by using the :cmake:variable:`RAPIDS_CMAKE_CPM_OVERRIDE_VERSION_FILE` set to the path of the generated pinned versions file.
+
+
+rapids-cmake cpm command line controls
+######################################
+
+rapids-cmake cpm offers multiple command line options to control behavior
+
+If you need to explicitly state a package must be downloaded and not searched
+for locally you enable the variable :cmake:variable:`CPM_DOWNLOAD_<package_name>`.
+
+```
+cmake -DCPM_DOWNLOAD_<package_name>=ON ....
```

Comment:
I would suggest a separate subheading that documents the most important underlying CPM options to delineate which options are ours and which come from CPM. Ideally we would link to relevant sections in CPM docs, but AFAIK the only docs are the CPM README and it's not that easy to link to sections, so it may just have to a be a single link to the file saying 'all CPM options can be found here'.

Reply:
Good idea. I have re-organized the section and would like a re-review.